### ISSUE TITLE
[Suggestion] Thermoregulation

### ISSUE THREAD

**Issue by Fokson** [Funwayguy/EnviroMine#153]
I do like Enviromine's concept very much, but as it is now I'm afraid I'm not willing to subject myself to the body temperature mechanic.  Humans have something called 'thermoregulation', meaning their bodies will automatically adjust (within reason) to external temperatures and keep their internal body temperature within acceptable limits.

In Minecraft with Enviromine, Steve functions more like a lizard, where his temperature is radically altered by any outside stimulus.  Standing two blocks away from a torch?  Expect hyperthermia within five minutes.  Raining in a comfortably warm biome?  Watch out, you'll be a block of ice in due time.

Perhaps this is what you intended, but in my opinion body temperature isn't something you should be constantly needing to babysit, minute to minute.  It should be dissuasion from dancing in your shorts in the snow, or wearing a parka in the desert.

My suggestion is to have your temperature taper towards 37 / 98.6 at all times, slowly, against whatever environmental factors there are.  If said factors outweigh your body's own ability to thermoregulate, then you're going to suffer their effects.  Additionally, sweating isn't just "A Thing That Happens When You're Hot"; sweating is a defense mechanism against high temperatures, and is supposed to lower the body's temperature even further.  I would suggest that sweating start at 38 Celsius as you have it, and cause the body's thermoregulation rate to increase.  Sweating should not stop until 37.5 Celsius.  Similarly, at 36 Celsius the body would technically begin to shiver, which is something the body does to keep warm.  As sweating lowers hydration, it would make sense for shivering to expend hunger.

Anyway, that's about all I have to say.  With that exception, Enviromine has been a very fun addition to my list of mods (my favorite bit being the physics)!

**Comment by GenDeathrow** [comment]
... funny you say this.. I have actually wanted to add in a more realistic thermoregulation for a long time, Ask funwayguy... I was actually wanting to write this go our rewrite.. but that could be some time.. the current system could do it.. but would require a major update. Im glad to hear others thinking of a similar system.

**Comment by Fokson** [comment]
I'm not fantastic at programming, but a long time ago I remember doing body-temperature related stuff in pseudocode with good results.  I might be able to lend a hand.

Could you inform me what the values in the array enviroData (EnviroDataTracker Line 120) represent?  Specifically, I'm not sure what [4] and [5] are, but it couldn't hurt to know the rest.

**Comment by GenDeathrow** [comment]
Currently im not at my computer.. on Mobile, but I will take a look at this when I do get a chance.. but I think you may have a very simple way of looking at this.. I think this will be much easier than I had originally thought.. I appreciate the info and if we do this Id like to maybe make some work in models with you for testing.. funwayguy of course will make all final decisions.

**Comment by Fokson** [comment]
Absolutely!  I'll be around, so let me know if you need anything else!

**Comment by GenDeathrow** [comment]
Ok, so if I am reading the Formula's right, the skin temp acts as a buffer between to the temps. Which would help with some of the crazy temp fluctuations that we can have now. Will have to tweak the rate of change because, we call the function about 20x per sec(based on update ticks). sweating does start at 38(unless some one changed it), but while sweating it prevents players from overheating while reducing hydration faster(I believe below 50% hydration players can overheat).  And I agree with reducing hunger while shivering... I have been wanting to add this as well.

**Comment by Fokson** [comment]
Yeah, I added the rate of change variable for two reasons.  One, as I wrote the formula on paper, I have no way of knowing exactly how fast it's going to be acting, and two, I recall a config option in Enviromine that allows you to adjust how quickly each stat works.

Skin temperature is partially a buffer, and partially just another touch of realism to handling body temperature like humans actually do.  The air doesn't directly effect our core temperature, it affects our skin, which then affects the inner body.  If you wanted to be really technical, you could have ambient item temperatures (in your inventory) affect skin temperature more directly, rather than having it go through the air temperature.  If you're holding a hot pack to your chest, it's going to outweigh the cold air.

Rather than using a constant for the thermoregulation target, you could make a variable that is set to 37, and when sweating it would be set lower (36, 35), and when shivering could be set higher (38, 39) so as to boost the body's ability to level out.

The reason I suggested that sweating / shivering have a different threshold for stopping than it does for starting is so it doesn't flicker on and off rapidly if it finds a ... "sweet spot" where it want to flicker between 38 and 37.9 rapidly.

**Comment by GenDeathrow** [comment]
I finally had a chance to use part of your code for testing, it actually looks like it may work much better than the current system. I'll keep ya updated. Thanks again,

**Comment by Funwayguy** [comment]
Just wanted to also thank you for the help with this. We've been trying to get this right for a long while without killing the CPU and it seems you've nailed it on the head with this. The algorithm is currently being implemented in the rewrite of EnviroMine (see the teardown branch) so it may not be in the current releases but it will be when the mod goes fully modular.

**Comment by Fokson** [comment]
You're very welcome!  I'm so glad some silly bit of pseudo-code I had lying around was able to help out so much! =)

If you need a hand with anything else, drop me a line.  I'm always happy to help!
